langchain/callbacks/promptlayer_callback.py

```diff
@@ -0,0 +1,135 @@
+import datetime
+from typing import Any, Callable, Dict, List, Optional, Tuple
+from uuid import UUID
+
+from langchain.callbacks.base import BaseCallbackHandler
+from langchain.schema import (
+    AIMessage,
+    BaseMessage,
+    ChatMessage,
+    HumanMessage,
+    LLMResult,
+    SystemMessage,
+)
+
+
+class PromptLayerCallbackHandler(BaseCallbackHandler):
+    def __init__(
+        self,
+        pl_id_callback: Optional[Callable[..., Any]] = None,
+        pl_tags: Optional[List[str]] = [],
+    ) -> None:
+        self.pl_id_callback = pl_id_callback
+        self.pl_tags = pl_tags
+
+        self.runs: Dict[UUID, Dict[str, Any]] = {}
+
+    def on_chat_model_start(
+        self,
+        serialized: Dict[str, Any],
+        messages: List[List[BaseMessage]],
+        *,
+        run_id: UUID,
+        parent_run_id: Optional[UUID] = None,
+        tags: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> Any:
+        self.runs[run_id] = {
+            "messages": [self._create_message_dicts(m)[0] for m in messages],
+            "invocation_params": kwargs.get("invocation_params", {}),
+            "name": kwargs.get("invocation_params", {}).get("_type", "No Type"),
+            "request_start_time": datetime.datetime.now().timestamp(),
+            "tags": tags,
+        }
+
+    def on_llm_start(
+        self,
+        serialized: Dict[str, Any],
+        prompts: List[str],
+        *,
+        run_id: UUID,
+        parent_run_id: Optional[UUID] = None,
+        tags: Optional[List[str]] = None,
+        **kwargs: Any,
+    ) -> Any:
+        self.runs[run_id] = {
+            "prompts": prompts,
+            "invocation_params": kwargs.get("invocation_params", {}),
+            "name": kwargs.get("invocation_params", {}).get("_type", "No Type"),
+            "request_start_time": datetime.datetime.now().timestamp(),
+            "tags": tags,
+        }
+
+    def on_llm_end(
+        self,
+        response: LLMResult,
+        *,
+        run_id: UUID,
+        parent_run_id: Optional[UUID] = None,
+        **kwargs: Any,
+    ) -> None:
+        run_info = self.runs.get(run_id, {})
+        run_info["request_end_time"] = datetime.datetime.now().timestamp()
+        if not run_info:
+            return
+        for i in range(len(response.generations)):
+            generation = response.generations[i][0]
+
+            resp = {
+                "text": generation.text,
+                "llm_output": response.llm_output,
+            }
+            model_params = run_info.get("invocation_params", {})
+            if run_info.get("name") == "openai-chat":
```

Comment:
If this is meant to infer whether it's a chat model or a base LLM, the more reliable way would be to check for th emessages key directly.

In its current state, this wouldn't handle things like Anthropic's chat messages